File: src/scripts/mixar/modules/mcp_bridge/blender/handlers/advanced/constraint.py
"""
Advanced constraint handlers for Blender MCP Bridge.
Provides: constraint/add, constraint/remove, constraint/configure,
          constraint/set-target, constraint/set-influence, constraint/list
"""

import bpy
from ...utils.response import ok_response, error_response, not_found, coerce_value
from ...utils.context_helpers import ensure_context_for_object, is_valid_rna_property
from .. import register_handler


# ─── Supported constraint types ─────────────────────────────────────────────────

_SUPPORTED_TYPES = {
    "COPY_LOCATION",
    "COPY_ROTATION",
    "COPY_SCALE",
    "TRACK_TO",
    "DAMPED_TRACK",
    "LIMIT_LOCATION",
    "LIMIT_ROTATION",
    "LIMIT_SCALE",
    "FOLLOW_PATH",
    "CLAMP_TO",
    "CHILD_OF",
    "FLOOR",
}


# ─── Helpers ─────────────────────────────────────────────────────────────────────

def _get_object(name):
    """Return (obj, None) or (None, error_response) for any object."""
    obj = bpy.data.objects.get(name)
    if obj is None:
        return None, not_found(name)
    return obj, None


def _get_constraint(obj, constraint_name):
    """Return (constraint, None) or (None, error_response)."""
    constraint = obj.constraints.get(constraint_name)
    if constraint is None:
        return None, error_response(
            f"Constraint '{constraint_name}' not found on '{obj.name}'."
        )
    return constraint, None


# ─── Handlers ────────────────────────────────────────────────────────────────────

def _handle_constraint_add(params):
    """
    Add a constraint to an object.
    Route: constraint/add
    """
    object_name = params.get("object_name")
    constraint_type = params.get("type")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not constraint_type:
        return error_response("Parameter 'type' is required.")
    if constraint_type not in _SUPPORTED_TYPES:
        return error_response(
            f"Unsupported constraint type '{constraint_type}'. "
            f"Valid types: {', '.join(sorted(_SUPPORTED_TYPES))}."
        )

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraint = obj.constraints.new(type=constraint_type)

        return ok_response({
            "object_name": obj.name,
            "constraint_name": constraint.name,
            "type": constraint_type,
        })
    except Exception as e:
        return error_response(f"Failed to add constraint: {e}")


def _handle_constraint_remove(params):
    """
    Remove a named constraint from an object.
    Route: constraint/remove
    """
    object_name = params.get("object_name")
    constraint_name = params.get("constraint_name")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not constraint_name:
        return error_response("Parameter 'constraint_name' is required.")

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraint = obj.constraints.get(constraint_name)
        if constraint is None:
            return error_response(
                f"Constraint '{constraint_name}' not found on '{object_name}'."
            )

        obj.constraints.remove(constraint)

        return ok_response({
            "object_name": object_name,
            "removed_constraint": constraint_name,
        })
    except Exception as e:
        return error_response(f"Failed to remove constraint: {e}")


def _handle_constraint_configure(params):
    """
    Set properties on an existing constraint via a params dictionary.
    Route: constraint/configure
    """
    object_name = params.get("object_name")
    constraint_name = params.get("constraint_name")
    configure_params = params.get("params")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not constraint_name:
        return error_response("Parameter 'constraint_name' is required.")
    if configure_params is None or not isinstance(configure_params, dict):
        return error_response("Parameter 'params' must be a non-empty object.")

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraint, err = _get_constraint(obj, constraint_name)
        if err:
            return err

        applied = []
        skipped = []

        for key, value in configure_params.items():
            if is_valid_rna_property(constraint, key):
                try:
                    current = getattr(constraint, key)
                    coerced = coerce_value(current, value)
                    setattr(constraint, key, coerced)
                    applied.append(key)
                except (AttributeError, TypeError, ValueError) as set_err:
                    skipped.append(f"{key} (set error: {set_err})")
            else:
                skipped.append(f"{key} (not a valid RNA property)")

        return ok_response({
            "object_name": object_name,
            "constraint_name": constraint_name,
            "type": constraint.type,
            "applied": applied,
            "skipped": skipped,
        })
    except Exception as e:
        return error_response(f"Failed to configure constraint: {e}")


def _handle_constraint_set_target(params):
    """
    Assign a target object (and optional target bone) to a constraint.
    Route: constraint/set-target
    """
    object_name = params.get("object_name")
    constraint_name = params.get("constraint_name")
    target_object = params.get("target_object")
    target_bone = params.get("target_bone")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not constraint_name:
        return error_response("Parameter 'constraint_name' is required.")
    if not target_object:
        return error_response("Parameter 'target_object' is required.")

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraint, err = _get_constraint(obj, constraint_name)
        if err:
            return err

        if not hasattr(constraint, "target"):
            return error_response(
                f"Constraint '{constraint_name}' (type: {constraint.type}) "
                "does not support a target object."
            )

        target = bpy.data.objects.get(target_object)
        if target is None:
            return not_found(target_object)

        constraint.target = target

        resolved_bone = None
        if target_bone:
            if hasattr(constraint, "subtarget"):
                constraint.subtarget = target_bone
                resolved_bone = target_bone
            else:
                # Not an error — just report it was skipped
                resolved_bone = None

        return ok_response({
            "object_name": object_name,
            "constraint_name": constraint_name,
            "type": constraint.type,
            "target_object": target.name,
            "target_bone": resolved_bone,
        })
    except Exception as e:
        return error_response(f"Failed to set constraint target: {e}")


def _handle_constraint_set_influence(params):
    """
    Set the influence value of a constraint (clamped to 0.0–1.0).
    Route: constraint/set-influence
    """
    object_name = params.get("object_name")
    constraint_name = params.get("constraint_name")
    influence = params.get("influence")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not constraint_name:
        return error_response("Parameter 'constraint_name' is required.")
    if influence is None:
        return error_response("Parameter 'influence' is required.")

    try:
        influence = float(influence)
    except (TypeError, ValueError):
        return error_response("Parameter 'influence' must be a number.")

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraint, err = _get_constraint(obj, constraint_name)
        if err:
            return err

        clamped = max(0.0, min(1.0, influence))
        constraint.influence = clamped

        return ok_response({
            "object_name": object_name,
            "constraint_name": constraint_name,
            "type": constraint.type,
            "influence": clamped,
        })
    except Exception as e:
        return error_response(f"Failed to set constraint influence: {e}")


def _handle_constraint_list(params):
    """
    List all constraints on an object.
    Route: constraint/list
    """
    object_name = params.get("object_name")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    try:
        obj = bpy.data.objects.get(object_name)
        if obj is None:
            return not_found(object_name)

        constraints = []
        for c in obj.constraints:
            info = {
                "name": c.name,
                "type": c.type,
                "influence": round(c.influence, 6),
                "mute": c.mute,
                "target": c.target.name if hasattr(c, "target") and c.target else None,
            }
            # Include subtarget (bone) if present and set
            if hasattr(c, "subtarget") and c.subtarget:
                info["target_bone"] = c.subtarget
            constraints.append(info)

        return ok_response({
            "object_name": obj.name,
            "constraints": constraints,
            "count": len(constraints),
        })
    except Exception as e:
        return error_response(f"Failed to list constraints: {e}")


# ─── Register routes ─────────────────────────────────────────────────────────────

register_handler("constraint", "add",           _handle_constraint_add)
register_handler("constraint", "remove",         _handle_constraint_remove)
register_handler("constraint", "configure",      _handle_constraint_configure)
register_handler("constraint", "set-target",     _handle_constraint_set_target)
register_handler("constraint", "set-influence",  _handle_constraint_set_influence)
register_handler("constraint", "list",           _handle_constraint_list)

File: mcp/src/blender/tools/advanced/constraint.js
// Advanced constraint tools — object constraints management

import { sendToBlender } from "../../blender-bridge.js";

export const constraintTools = [
  {
    name: "blender_constraint_add",
    description:
      "Add a constraint to an object. " +
      "Supported types: COPY_LOCATION (copy another object's location), " +
      "COPY_ROTATION (copy another object's rotation), " +
      "COPY_SCALE (copy another object's scale), " +
      "TRACK_TO (make the object point toward a target), " +
      "DAMPED_TRACK (damped tracking toward a target axis), " +
      "LIMIT_LOCATION (restrict translation within min/max bounds), " +
      "LIMIT_ROTATION (restrict rotation within min/max bounds), " +
      "LIMIT_SCALE (restrict scale within min/max bounds), " +
      "FOLLOW_PATH (move the object along a curve path), " +
      "CLAMP_TO (clamp object position to a curve), " +
      "CHILD_OF (make the object behave as a child of a target without re-parenting), " +
      "FLOOR (prevent the object from passing through a target surface). " +
      "Returns the assigned constraint name (Blender may append a suffix for uniqueness). " +
      "Example call: { \"object_name\": \"Cube\", \"type\": \"COPY_LOCATION\" }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object to add the constraint to.",
        },
        type: {
          type: "string",
          enum: [
            "COPY_LOCATION",
            "COPY_ROTATION",
            "COPY_SCALE",
            "TRACK_TO",
            "DAMPED_TRACK",
            "LIMIT_LOCATION",
            "LIMIT_ROTATION",
            "LIMIT_SCALE",
            "FOLLOW_PATH",
            "CLAMP_TO",
            "CHILD_OF",
            "FLOOR",
          ],
          description: "The constraint type identifier. Must be UPPERCASE.",
        },
      },
      required: ["object_name", "type"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_constraint_remove",
    description:
      "Remove a named constraint from an object. " +
      "Use blender_constraint_list to retrieve the exact constraint name before calling this. " +
      "Returns the object name and the removed constraint name. " +
      "Example call: { \"object_name\": \"Cube\", \"constraint_name\": \"Copy Location\" }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object that owns the constraint.",
        },
        constraint_name: {
          type: "string",
          description:
            "Exact name of the constraint to remove (as returned by blender_constraint_list or blender_constraint_add).",
        },
      },
      required: ["object_name", "constraint_name"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/remove", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_constraint_configure",
    description:
      "Set one or more properties on an existing constraint via a params dictionary. " +
      "Only properties that exist on the constraint type are applied; unrecognised keys are skipped. " +
      "Common params by constraint type: " +
      "COPY_LOCATION/ROTATION/SCALE — use_x, use_y, use_z (bool), invert_x, invert_y, invert_z (bool), use_offset (bool); " +
      "TRACK_TO — track_axis ('TRACK_X','TRACK_Y','TRACK_Z','TRACK_NEGATIVE_X','TRACK_NEGATIVE_Y','TRACK_NEGATIVE_Z'), up_axis ('UP_X','UP_Y','UP_Z'); " +
      "LIMIT_LOCATION — use_min_x, min_x, use_max_x, max_x, use_min_y, min_y, use_max_y, max_y, use_min_z, min_z, use_max_z, max_z (floats/bools); " +
      "LIMIT_ROTATION — use_limit_x, min_x, max_x, use_limit_y, min_y, max_y, use_limit_z, min_z, max_z (degrees as floats); " +
      "FOLLOW_PATH — forward_axis ('FORWARD_X','FORWARD_Y','FORWARD_Z'), up_axis, use_fixed_location (bool), offset (float); " +
      "FLOOR — floor_location ('FLOOR_X','FLOOR_Y','FLOOR_Z','FLOOR_NEGATIVE_X','FLOOR_NEGATIVE_Y','FLOOR_NEGATIVE_Z'), offset (float), use_sticky (bool). " +
      "Returns a list of applied keys and any skipped keys. " +
      "Example call: { \"object_name\": \"Cube\", \"constraint_name\": \"Copy Location\", \"params\": { \"use_x\": true, \"use_y\": false, \"influence\": 0.5 } }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object that owns the constraint.",
        },
        constraint_name: {
          type: "string",
          description: "Name of the constraint to configure.",
        },
        params: {
          type: "object",
          description:
            "Key/value pairs to apply to the constraint. " +
            "IMPORTANT: Must be a JSON object, NOT a string. " +
            "Each key must match a valid Python attribute name on the constraint type. " +
            "Example: { \"use_x\": true, \"use_y\": false, \"influence\": 0.5 }.",
        },
      },
      required: ["object_name", "constraint_name", "params"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/configure", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_constraint_set_target",
    description:
      "Assign a target object (and optionally a target bone) to a constraint. " +
      "Applies to constraints that have a 'target' property such as COPY_LOCATION, COPY_ROTATION, " +
      "COPY_SCALE, TRACK_TO, DAMPED_TRACK, FOLLOW_PATH, CLAMP_TO, CHILD_OF, and FLOOR. " +
      "When targeting a specific bone on an armature, provide target_bone with the pose bone name. " +
      "Returns the object name, constraint name, resolved target object name, and target bone (if set). " +
      "Example call: { \"object_name\": \"Cube\", \"constraint_name\": \"Copy Location\", \"target_object\": \"Empty\" }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the constrained object.",
        },
        constraint_name: {
          type: "string",
          description: "Name of the constraint to update.",
        },
        target_object: {
          type: "string",
          description: "Name of the Blender object to use as the constraint target.",
        },
        target_bone: {
          type: "string",
          description:
            "Optional. Pose bone name when the target object is an armature and you want to track a specific bone.",
        },
      },
      required: ["object_name", "constraint_name", "target_object"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/set-target", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_constraint_set_influence",
    description:
      "Set the influence (blend weight) of a constraint, controlling how strongly it affects the object. " +
      "A value of 1.0 means full effect; 0.0 means no effect. " +
      "The value is automatically clamped to the [0.0, 1.0] range. " +
      "Returns the object name, constraint name, and the applied influence value. " +
      "Example call: { \"object_name\": \"Cube\", \"constraint_name\": \"Copy Location\", \"influence\": 0.5 }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the constrained object.",
        },
        constraint_name: {
          type: "string",
          description: "Name of the constraint to adjust.",
        },
        influence: {
          type: "number",
          minimum: 0,
          maximum: 1,
          description: "Influence value in the range 0.0 (no effect) to 1.0 (full effect).",
        },
      },
      required: ["object_name", "constraint_name", "influence"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/set-influence", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_constraint_list",
    description:
      "List all constraints on an object, returning their names, types, influence values, " +
      "mute state, and target object name (if any). " +
      "Use this to discover constraint names before calling configure, set-target, set-influence, or remove. " +
      "Example call: { \"object_name\": \"Cube\" }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object whose constraints to list.",
        },
      },
      required: ["object_name"],
    },
    handler: async (args) => {
      const result = await sendToBlender("constraint/list", args);
      return JSON.stringify(result, null, 2);
    },
  },
];
